[#27] Redis를 세션 스토리지로 활용하여 로그인 구현하기

sgo8308 opened the pull request:
### **개요**

세션 스토리지로 Redis를 활용하여 로그인 기능을 구현했습니다.

### **이슈 링크**

#27 
### **작업 사항 (변경 사항)**
* 로그인하게 될 경우 로그인 유저의 id와 이름을 담고 있는 LoginUser 객체는 Json 방식으로 직렬화되어 Redis에 저장됩니다.
* 보안 문제와 직렬화 데이터 용량 문제로 인해 자바 직렬화 방식이 아닌 Json 직렬화 방식을 사용했습니다.

### 테스트 내역
사전 준비 사항
- 로컬 MySQL user 테이블에 유저 데이터 하나 넣기
- 로컬 VMWare에 Redis를 설치

테스트
* Postman을 통하여 /login uri로 로그인을 시도한 후 LoginUser 객체가 Json 형태로 저장되는 것을 확인했습니다. 
* 로그인에 성공한 후 /users/me로 내 정보 보기를 시도했을 때 정상적으로 정보들이 응답 데이터로 전달되는 것을 확인했습니다. 

### **리뷰 받고 싶은 내용**

sgo8308 committed:
feat(user): redis 를 세션 스토리지로 사용하도록 변경

현재는 구조로는 스케일 아웃시 서버 간 세션 공유가 안되기 때문에 레디스를 사용하도록 변경

Closes #27

sgo8308 committed:
Merge remote-tracking branch 'origin/develop' into feature/27

sgo8308 committed:
refactor(user): RedisSession이 직렬화 시 자바 직렬화가 아닌 Json 직렬화를 사용하도록 변경

자바 직렬화는 보안상 문제가 있을 수 있으며 직렬화된 객체의 용량 또한 Json 방식보다 최소 2배가 크다
Json 직렬화 방식 중 타입을 명확히 지정하기 때문에 더 적은 용량으로 저장할 수 있는
Jackson2JsonRedisSerializer를 GenericJackson2JsonRedisSerializer 대신에 사용

sgo8308 committed:
refactor(common): SessionConstant.LOGIN_KEY를 "name"에서 "loginUser"로 변경

derek-Flab commented on src/main/java/com/flab/marketgola/user/config/RedisConfig.java:
```diff
@@ -0,0 +1,18 @@
+    @Bean
+    @Qualifier("springSessionDefaultRedisSerializer")
+    public RedisSerializer<Object> springSessionDefaultRedisSerializer() {
+        return new Jackson2JsonRedisSerializer(LoginUser.class);
```
1. 제네릭 클래스이므로 암시적으로 생략이 가능한 `Jackson2JsonRedisSerializer <>(..)`이렇게 선언해주는게 좋겠습니다
2. 빈 메서드의 리턴 타입이 `RedisSerializer<Object>`이던데 제네릭 타입을 Object로 선언해야지만 동작하나요?LoginUser 타입으로 리턴해도 되는 것 아닌지?

sgo8308 replied:
RedisSession을 이용해서 관련 작업을 처리하는 RedisIndexedSessionRepository의 defaultSerializer가 다음과 같이 setter주입 방식을 이용해서 주입됩니다. 이 때 타입이 RedisSerializer<Object>이기 때문에 RedisSerializer<Object>라고 RedisConfig 파일에 선언해주어야지 제가 만들어준 DefaultRedisSerializer로 대체되는 것 같습니다. 또 Jackson2JsonRedisSerializer <>(LoginUser.class)와 같이 표현할 경우 실제로 Jackson2JsonRedisSerializer<LoginUser>(LoginUser.class);와 같아지기 때문에 RedisSerializer<Object>과 호환이 안되서 컴파일 오류가 나게 됩니다 ! 

```
@Autowired(required = false)
	@Qualifier("springSessionDefaultRedisSerializer")
	public void setDefaultRedisSerializer(RedisSerializer<Object> defaultRedisSerializer) {
		this.defaultRedisSerializer = defaultRedisSerializer;
	}
```

derek-Flab reviewed (commented):
엇. 제가 코드 리뷰를 깜빡했네요! 
죄송요
크게 리뷰드릴 내용은 없고, 리뷰에 대해서 지우님이 최종 확인 해주세요~. 

아 그리고 CI가 깨지는데 이 부분은 수정해서 최종 CI통과까지 확인해주시는게 좋겠어요!

sgo8308 committed:
ci: CI Script에 Redis 관련 환경 변수 추가

derek-Flab commented:
CI실패가 났네요!? 확인해주셔야 될 것 같습니다!

sgo8308 committed:
ci: CI 스크립트에서 Redis 관련 환경 변수 제거

CI 상황에서는 Embedded-Redis를 사용하기로 변경했기 때문에

sgo8308 committed:
feat: embedded redis 기능 추가 후 관련 테스트들 수정

CI 상황에서는 Embedded-Redis를 사용하여 테스트
실제 환경과 유사한 상황의 테스트는 CI 후 빌드된 결과물을 가지고 dev 서버에서 진행
